### PR TITLE
CK: Add Switchboard Feeds

### PR DESCRIPTION
**Pull request by coachchucksol**
- Added Feeds to permissionlessly set weights
- Changed TrackedMints -> VaultRegistry
- Renamed some instructions to identify what needs an admin

### CODE REVIEW

File: program/src/snapshot_vault_operator_delegation.rs [hunk 1]
```diff
@@ -111,13 +109,13 @@ pub fn process_snapshot_vault_operator_delegation(
...
-        let mint_entry = tracked_mints_account.get_mint_entry(vault_index)?;
+        let weight_table_data = weight_table.data.borrow();
+        let weight_table_account = WeightTable::try_from_slice_unchecked(&weight_table_data)?;
+        let weight_entry = weight_table_account.get_weight_entry(&st_mint)?;
 
         (
-            mint_entry.ncn_fee_group(),
-            mint_entry.reward_multiplier_bps(),
+            weight_entry.mint_entry().ncn_fee_group(),
+            weight_entry.mint_entry().reward_multiplier_bps(),
```
Comment (ebatsell) [inline program/src/snapshot_vault_operator_delegation.rs:118]: weight table account is also deserialized below, should consolidate so we only deser once

File: program/src/switchboard_set_weight.rs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        let weight_entry = weight_table_account.get_weight_entry(&st_mint)?;
+
+        (
+            weight_entry.mint_entry().switchboard_feed(),
+            weight_entry.mint_entry().no_feed_weight(),
+        )
+    };
+
+    let weight: u128 = if registered_switchboard_feed.eq(&Pubkey::default()) {
+        no_feed_weight
```
Comment (ebatsell) [inline program/src/switchboard_set_weight.rs:45]: maybe a msg! here that we are going with fallback weight. a bit worried about scenarios where fallback weight is being used when we thought there was a feed or forgot to add a feed

File: program/src/initialize_epoch_snapshot.rs [hunk 3]
```diff
@@ -96,9 +95,9 @@ pub fn process_initialize_epoch_snapshot(
     };
 
     let vault_count: u64 = {
-        let tracked_mints_data = tracked_mints.data.borrow();
-        let tracked_mints_account = TrackedMints::try_from_slice_unchecked(&tracked_mints_data)?;
-        tracked_mints_account.mint_count()
+        let weight_table_data = weight_table.data.borrow();
+        let weight_table_account = WeightTable::try_from_slice_unchecked(&weight_table_data)?;
+        weight_table_account.vault_count()
```
Comment (ebatsell) [inline program/src/initialize_epoch_snapshot.rs:100]: consolidate with other weight table deserialize above

File: core/src/vault_registry.rs [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+        if account.data.borrow()[0].ne(&Self::DISCRIMINATOR) {
+            msg!("Tracked Mints account discriminator is invalid");
+            return Err(ProgramError::InvalidAccountData);
+        }
+
+        if account
+            .key
+            .ne(&Self::find_program_address(program_id, ncn).0)
+        {
+            msg!("Tracked Mints account is not at the correct PDA");
```
Comment (ebatsell) [inline core/src/vault_registry.rs:197]: Update messages from Tracked Mints name

File: core/src/vault_registry.rs [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+impl VaultRegistry {
+    pub const SIZE: usize = 8 + size_of::<Self>();
+
+    pub fn new(ncn: Pubkey, bump: u8) -> Self {
+        Self {
+            ncn,
+            bump,
+            reserved: [0; 127],
+            st_mint_list: [StMintEntry::default(); MAX_ST_MINTS],
+            vault_list: [VaultEntry::default(); MAX_VAULT_OPERATOR_DELEGATIONS],
```
Comment (ebatsell) [inline core/src/vault_registry.rs:139]: do you remember why we called this `MAX_VAULT_OPERATOR_DELEGATIONS` ? should just be `MAX_VAULTS` right?

File: core/src/vault_registry.rs [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+
+#[derive(Debug, Clone, Copy, Zeroable, ShankType, Pod)]
+#[repr(C)]
+pub struct StMintEntry {
+    st_mint: Pubkey,
+    ncn_fee_group: NcnFeeGroup,
+    reward_multiplier_bps: PodU64,
+    switchboard_feed: Pubkey,
+    no_feed_weight: PodU128,
+    reserved: [u8; 32],
```
Comment (ebatsell) [inline core/src/vault_registry.rs:26]: would add more reserve space here, could see that being useful for other feed ids or metadata

File: core/src/vault_registry.rs [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+        let tracked_mints = VaultRegistry::new(Pubkey::default(), 0);
+        assert_eq!(
+            tracked_mints.vault_list.len(),
+            MAX_VAULT_OPERATOR_DELEGATIONS
+        );
+    }
+
+    #[test]
+    fn test_add_mint() {
+        let mut tracked_mints = VaultRegistry::new(Pubkey::default(), 0);
```
Comment (ebatsell) [inline core/src/vault_registry.rs:372]: tracked mints in a lot of these test names

File: core/src/vault_registry.rs [hunk 8]
```diff
@@ -0,0 +1,433 @@
...
+        {
+            return Err(TipRouterError::VaultIndexAlreadyInUse.into());
+        }
+
+        // Insert at the first empty slot
+        let mint_entry = self
+            .vault_list
+            .iter_mut()
+            .find(|m| m.st_mint == VaultEntry::default().st_mint)
+            .ok_or(TipRouterError::TrackedMintListFull)?;
```
Comment (ebatsell) [inline core/src/vault_registry.rs:310]: Update error name

File: integration_tests/tests/fixtures/tip_router_client.rs [hunk 9]
```diff
@@ -591,49 +671,165 @@ impl TipRouterClient {
...
+        let restaking_config = Config::find_program_address(&jito_restaking_program::id()).0;
+
+        let vault_registry =
+            VaultRegistry::find_program_address(&jito_tip_router_program::id(), &ncn).0;
+
+        let (ncn_config, _, _) =
+            NcnConfig::find_program_address(&jito_tip_router_program::id(), &ncn);
+
+        //TODO: Check admin is correct
+        let admin = self.payer.pubkey();
```
Comment (ebatsell) [inline integration_tests/tests/fixtures/tip_router_client.rs:692]: todo?

File: program/src/register_vault.rs [hunk 10]
```diff
@@ -80,14 +59,18 @@ pub fn process_register_mint(program_id: &Pubkey, accounts: &[AccountInfo]) -> P
         return Err(ProgramError::InvalidAccountData);
     }
 
+    let mut tracked_mints_data = vault_registry.try_borrow_mut_data()?;
+    let tracked_mints = VaultRegistry::try_from_slice_unchecked_mut(&mut tracked_mints_data)?;
```
Comment (ebatsell) [inline program/src/register_vault.rs:63]: tracked mints

File: core/src/vault_registry.rs (hunk 8 above)
Comment (coachchucksol) [reply core/src/vault_registry.rs:310]: done

File: program/src/snapshot_vault_operator_delegation.rs (hunk 1 above)
Comment (coachchucksol) [reply program/src/snapshot_vault_operator_delegation.rs:118]: Done

File: core/src/vault_registry.rs (hunk 5 above)
Comment (coachchucksol) [reply core/src/vault_registry.rs:139]: Correct, it should be MAX_VAULTS.

File: core/src/vault_registry.rs (hunk 7 above)
Comment (coachchucksol) [reply core/src/vault_registry.rs:372]: should be addressed

File: core/src/vault_registry.rs (hunk 4 above)
Comment (coachchucksol) [reply core/src/vault_registry.rs:197]: Done

File: program/src/initialize_weight_table.rs [hunk 11]
```diff
@@ -77,8 +77,34 @@ pub fn process_initialize_weight_table(
...
+    // let weight_table_account = WeightTable::try_from_slice_unchecked_mut(&mut weight_table_data)?;
+
+    // weight_table_account.initialize(
+    //     *ncn.key,
+    //     epoch,
+    //     Clock::get()?.slot,
+    //     vault_count,
+    //     weight_table_bump,
+    //     &mint_entries,
+    // )?;
```
[the remark this answers is not in the document]
Comment (coachchucksol) [reply program/src/initialize_weight_table.rs:107]: Done

File: program/src/initialize_weight_table.rs [hunk 12]
```diff
@@ -46,8 +46,8 @@ pub fn process_initialize_weight_table(
 
     let tracked_mint_count = {
         let tracked_mints_data = tracked_mints.data.borrow();
-        let tracked_mints = TrackedMints::try_from_slice_unchecked(&tracked_mints_data)?;
-        tracked_mints.mint_count()
+        let tracked_mints = VaultRegistry::try_from_slice_unchecked(&tracked_mints_data)?;
+        tracked_mints.vault_count()
     };
```
[the remark this answers is not in the document]
Comment (coachchucksol) [reply program/src/initialize_weight_table.rs:51]: Done

File: integration_tests/tests/fixtures/tip_router_client.rs (hunk 9 above)
Comment (coachchucksol) [reply integration_tests/tests/fixtures/tip_router_client.rs:692]: Removing

File: program/src/initialize_vault_registry.rs [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+    create_account(
+        payer,
+        vault_registry,
+        system_program,
+        program_id,
+        &Rent::get()?,
+        MAX_REALLOC_BYTES,
+        // 8_u64
+        //     .checked_add(std::mem::size_of::<VaultRegistry>() as u64)
+        //     .unwrap(),
```
[the remark this answers is not in the document]
Comment (coachchucksol) [reply program/src/initialize_vault_registry.rs:44]: Done

File: program/src/register_vault.rs (hunk 10 above)
Comment (coachchucksol) [reply program/src/register_vault.rs]: Done

File: program/src/initialize_vault_registry.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (coachchucksol) [reply program/src/initialize_vault_registry.rs]: Done ( I forgot to save/push this )

**Review by ebatsell — APPROVED** [review]

**Pull request merged**